On this project we're working as data analysts for a fictional company that builds Android and iOS mobile apps. We make our apps available on Google Play and in the App Store. 

We only build apps that are free to download and install, and our main source of revenue consists of in-app ads. This means that the number of users of our apps determines our revenue for any given app — the more users who see and engage with the ads, the better. 

Our goal for this project is to analyze data to help our developers understand what type of apps are likely to attract more users.


In [1]:
from csv import reader
opened_file = open('googleplaystore.csv')
read_file = reader(opened_file)
android = list(read_file)
android_header = android[0]
android = android[1:]

opened_file = open('AppleStore.csv')
read_file = reader(opened_file)
iOS = list(read_file)
iOS_header = iOS[0]
iOS = iOS[1:]

The first thing we need to do is to open both datasets, taken [from](https://www.kaggle.com/datasets/lava18/google-play-store-apps)[here](https://www.kaggle.com/datasets/ramamet4/app-store-apple-data-set-10k-apps), and create variables for both the main dataset which contains all the main information and another one just for the headers. Why? Because with just one single *print()* we can easily identify which column or row we need to extract the data from.

In [2]:
def explore_dataset(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]
    for row in dataset_slice:
        print(row)
        print('\n')
    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns', len(dataset[0]))

Then, we've created a new dataset named *explore_dataset* which contains *dataset*, a list of lists, the *start* and the *end* of the data, and a boolean entitled *rows_and_columns*. To establish the limits of the dataset, we have the variable *dataset_slice*, which marks the beginning and the end of it. Then looped the *rows* in that variable, and we've added an IF conditional on *rows_and_columns* that will print the number of rows and columns that each dataset has. Its length, in other words. 

In [3]:
print(android_header)
print('\n')
explore_dataset(android, 0, 3, True)

print(iOS_header)
print('\n')
explore_dataset(iOS, 0, 3, True)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


Number of rows: 10841
Number of columns 13
['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215'

Thanks to our new function and variables we can easily know how many rows and columns we have on android and iOS. This is key to know the starting point, as we will need to check out and filter the data and what we know might differ from what we will have at the end of this project. 

Cleaning data:

In [4]:
print(android[10472])
print('\n')
print(android_header)
print('\n')
print(android[0])

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


In [5]:
print(len(android))
del android[10472]
print(len(android))

10841
10840


Our next step is practicing cleaning data. For that, in the Google Play Store forum,there's a [discussion](https://www.kaggle.com/datasets/lava18/google-play-store-apps/discussion/66015) about an error located in the row **10472**. To know why this row gives us error, we're going to print that one, the header, and another row that doesn't have any error.

First row is the **10472**, which corresponds to the App *‘Life Made WiFi Touchscreen Photo Frame’*. 
Second row is the header of the entire dataset, in other words, which categories we can find in the columns below.
Third row is a row that presents no further issues

What's the error, if everything seems normal? We have to take a look at the **print(android_header)** row first, the third column would correspond with *‘Rating’*, but in the third column of the **print(android[10472])** we can see that it would correspond with *‘19’*, since in the Google Play Store the ratings just go up as high as 5 stars, best case scenario this column should reflect a number from 1 to 5, not 19, which indicates to us that some data might be missing. 

If we search in the [forum](https://www.kaggle.com/datasets/lava18/google-play-store-apps/discussion/66015), many users point out that there’s a missing value in the column data named *‘Category’*, therefore the column values shift and that’s why we had that weird result. Even the creator of the dataset confirms that 1.9 is the app rating. 

In order to fix this disorder, we have to eliminate the row, for the remaining rows to match the headers. 

For that we've used the **del** statement, and now we have **10840** rows instead of the previous **10841** we had. 

After further checking out both the Google Play dataset and the discussion, we've noticed that there are some duplicate apps, which means that from all the 10840 rows this dataset currently has, not all of them repressent unique values. For example, there are various rows dedicated to Social Media Platforms

In [6]:
for app in android:
    name = app[0]
    if name == 'Instagram':
        print('\n')
        print(app)
        
for app in android:
    name = app[0]
    if name == 'Facebook':
        print('\n')
        print(app)
        
for app in android:
    name = app[0]
    if name == 'Twitter':
        print('\n')
        print(app)



['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


['Facebook', 'SOCIAL', '4.1', '78158306', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'August 3, 2018', 'Varies with device', 'Varies with device']


['Facebook', 'SOCIAL', '4.1', '78128208', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 

As we can observe, with just a quick look we can see that there are **four rows** dedicated to 'Instagram', **two rows** dedicated to 'Faceebok' and **three rows** dedicated to 'Twitter', with almost identical data, which raises the following question: Which one should we eliminate?

The key is the **fourth column** on each row, which is the only data that differs from one row to another. What repressent the fourth column, then? We just have to **print** again the **headers of the android dataset** to know it.

In [7]:
print(android_header)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


The discordant column corresponds to *Reviews* which means that the app version with the most reviews will be the most updated. But how many more duplicate apps will we find beyond the social media ones? 

In [8]:
duplicate_apps = []
unique_apps = []

for app in android:
    name = app[0]
    if name in unique_apps:
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)
print('Number of duplicate apps:', len(duplicate_apps))
print('\n')
print('Number of unique apps:', len(unique_apps))
print('\n')
print('Examples of duplicate apps:', duplicate_apps[:10])
print('\n')
print('Examples of unique apps:', unique_apps[:10])

Number of duplicate apps: 1181


Number of unique apps: 9659


Examples of duplicate apps: ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack']


Examples of unique apps: ['Photo Editor & Candy Camera & Grid & ScrapBook', 'Coloring book moana', 'U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'Sketch - Draw & Paint', 'Pixel Draw - Number Art Coloring Book', 'Paper flowers instructions', 'Smoke Effect Photo Maker - Smoke Editor', 'Infinite Painter', 'Garden Coloring Book', 'Kids Paint Free - Drawing Fun']


To solve how many repeated apps we have, we've created **two different lists**, one for **duplicate apps**, another one for **unique apps**. 

Then, we've looped the apps in android by creating a new variable called *'name'* and introduced an **IF** condition. If the name of the app appears more than once in the unique apps list (that's why we've looped the apps), it means that it should go to the duplicate list. In other words, we will *append it* to that list
If not (the *else* statement) and it just appears once, it will remain on the unique apps list. 

At last we've printed the number of both duplicate and unique apps, together with a few examples to get the exact number of how many rows we have to eliminate for having repeated apps: **1181**. 

But between identical apps, which will erase and which will survive? Based on the *'Reviews'* criteria we will just leave the versions with the highest number of reviews, as we understand that they are the most updated versions. 

In [9]:
reviews_max = {}

for app in android:
    name = app[0]
    n_reviews = float(app[3])
    
    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews 
        
    elif name not in reviews_max:
        reviews_max[name] = n_reviews
        
print('Actual length', len(android) - 1181) 
print('Expected length', len(reviews_max))

Actual length 9659
Expected length 9659


Our next target is to build a dictionary where we will use as keys unique values with the name of the apps, whose values will be the nº of reviews. For that, we've named it **reviews_max**

We've looped the app row in the google play store dataset (android), with *name* taking data from *app*, which we've just looped, then, we've converted to **float** the data from the reviews of the apps (hence the app[3]) and put it under the function *n_reviews*. Also, we've introduced a couple of conditions: If *name* (making reference to the name column of the dataset) is already a key in our new dictionary and the number of reviews_max is **less than** n_reviews, then we will update the former.

On the contrary, if name is not in reviews_max, it will automatically create a new entry in the dictionary with a unique app as a key and the value is the number of reviews. We have used elif to tell python from doing more operations. 

Once we've done the first calculations, we've printed the length of the android dataset minus the 1181 reviews, and the result is 9659, which matches with the expected length and the calculations we did in the exercise before. 

In [10]:
android_clean = []
already_added = []

for app in android:
    name = app[0]
    n_reviews = float(app[3])
    
    if (n_reviews == reviews_max[name]) and (name not in already_added):
            android_clean.append(app)
            already_added.append(name)

We've created two empty lists, **android_clean**, where we will try to put all of out cleansed data, and the **already_added list**, with just app names, unclean data if you please. 

Again, we've looped the entire row (under the name of *app*) on the **android** dataset, containing the entirety of the app names (hence the 0) under the variable *name*, and converted the number of reviews (the app[3]) to float and put it under the variable *n_reviews*

Then it's time to introduce a couple of **conditions**: If the *n_reviews* equals to the *max number* of reviews of the app *name* (again, making reference to *name column* of the dataset), **and** that *name* is not in the already_added list, we will have to append data to both of the list we've created both. It's worth mention that both conditions are mandatory in order to append. 

Once conditions are met, the entire *app* row will be appended to the **android_clean** list, and the *name* of the app will go to the **already_added** list, where we will have the app names, as we stated before. 







In [11]:
explore_dataset(android_clean, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9659
Number of columns 13


As expected, after using our created function **explore_dataset** on the *android_clean* list, the *0*, and *3* rows and checking if everything was **True**, we got the same number of rows, **9659** and **13** columns. 

Since our company develop apps on English since this is the language of our audience, as we did with the duplicate apps, we need to remove from the database those apps not destined for an English-speaking market. Since we have almost 10.000 rows, there's a high chance that we encounter some apps in another languages. For example: 

In [12]:
print(android_clean[4412][0])
print(android_clean[7940][0])

中国語 AQリスニング
لعبة تقدر تربح DZ


We're going to do the following: Our rule for erasing apps will be based on symbols. We will eliminate those which doesn't use symbols commonly used in English test, which ranges from letters from the alphabet (A-Z), numbers composed composed of digits from 0 to 9, punctuation marks (., !, ?, ;), and other symbols (+, *, /).* Those characters have a numeric equivalence in Python. For example, 'a' corresponds to 97, while 'A' to 65. We know that thanks to the build-in function ord(). 

The numbers corresponding to the characters we commonly use in an English text are all in the range **0 to 127**, which means that anything that falls beyond that rank, won't be an English character. 

Based on that, we'll create a function that will rank all characters between 0-127 as **True** and those above as **False**

In [13]:
def english_apps(string):
    
    for character in string:
        if ord(character) > 127:
            return False
        
    return True

print(english_apps('Instagram'))
print(english_apps('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(english_apps('Docs To Go™ Free Office Suite'))
print(english_apps('Instachat 😜'))

True
False
False
False


In order for our function **english_apps** to properly work, which iterates over the imput *string*, we've looped the *character* in the *string*, then used the ord() function to know the number of each character. If that character **is bigger than 127** it will return as **False**, if it doesn't meet that condition will simply return as **True**. 

We've tried four examples to test our brand new function, but more problems arise. With the first two, it correctly analyzes if the characters are True (between 0-127) or False (greater than 127), but the following two uses emojis and special symbols that can't be detected by the normal code system, therefore the results are deceitful. 

Those two are English apps, but because of the special symbols, Python doesn't detect them as such. And we could be eliminating worthy apps. So we have to think about a solution for the function.  

One plausible solution could be eliminate an app if it has more than three emojis and/or special characters. While it's not a perfect solution  some functions might still be unfairly erased, it should work. 

In [14]:
def english_apps(string):
    non_ascii = 0
    
    for character in string:
        if ord(character) > 127:
            non_ascii += 1
        
    if non_ascii > 3:
        return False
    else:
        return True

print(english_apps('Docs To Go™ Free Office Suite'))
print(english_apps('Instachat 😜'))
print(english_apps('爱奇艺PPS -《欢乐颂2》电视剧热播'))

True
True
False


We made a few modifications: First of we've introduced together with the new function **english_apps** the variable *non_ascii* as empty or **0**, which is based on the **ASCII system**, the one which determines the range 0 to 127.

Then we've looped again the *character*, and it has to have a few conditions: If is superior to 127 we will sum 1 to the value of *non_ascii*, and if the value of *non_ascii* is **greater than 3**, then the result should be **False**. If those conditions doesn't exist, the result will be **True**. 

After printing again the two apps with special symbols, we see that this time Python detects them as True, thanks to the second conditional of *if non_ascii > 3*

After checking out that the function works properly, it's time to test in on both datasets for Google Play and iOS. 

In [15]:
print(android_header)
print('\n')
print(iOS_header)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


In [16]:
android_english = []
iOS_english = []

for app in android_clean:
    name = app[0]
    if english_apps(name):
        android_english.append(app)
        
for app in iOS:
    name = app[1]
    if english_apps(name):
        iOS_english.append(app)
        
explore_dataset(android_english, 0, 3, True)
print('\n')
explore_dataset(iOS_english, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9614
Number of columns 13


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', '

We've created two empty lists, one for each dataset. 

First we've looped the *app* in our *android_clean* and *iOS* datasets and created the variable *name*, which will contain the names of the apps. If the android and iOS databases have [0] and [1] respectively it is because in the datasets, the columns in the datasets have a different order, being *name* the first column in Google Play Store, while being the second in the iOS.

Then, we've used our new function *english_apps* on the *name*. If that *name* meets the criteria for the english apps set in our previous function (< 127  or non_ascii < 3) it will be appended to the lists **android_english/iOS_english**. 

Finally we've used the *explore_datasets* function to know exactly how many english apps we have on the datasets. Which gives us **9614** on Android and **6183** on iOS.

Since our company just creates apps that are free to download and install and we make our money through in-app ads, we need to separate free and payment apps from the datasets we've been working on and so far we've eliminated the incorrect data, the duplicate apps and the non-english apps. For that, our next step is to isolate free apps. 

In [17]:
android_free = []
iOS_free = []

for app in android_english:
    price = app[7]
    if price == '0':
        android_free.append(app)

for app in iOS_english:
    price = app[4]
    if price == '0.0':
        iOS_free.append(app)
        
print("Number of rows in Android:", len(android_free))
print("Number of rows in iOS:", len(iOS_free))

Number of rows in Android: 8864
Number of rows in iOS: 3222


The easiest way to isolate those free apps is to create separate lists for them, android and iOS free.

Then looped our already cleansed datasets from the previous step, created a variable named *price* instead of the usual *name* so far, which will take data from their respective *price* columns in the android and iOS datasets (hence the 7 and 4, as each has their own order). 

Then we've introduced an IF conditional, but this one is tricky and easy to miss. Since prices come up as strings in the datasets ('0', $0.99, $6.66, etc.) we can't just work with integers or floats in this part, since something like price == 0 will confuse Python and will make it believe that *price* is equivalent to 0, therefore the result of the rows will always be **0**, no matter what. That's why those zeros are put as strings. 

Following the normal course of action, we then tell Python that if *price* of an app in the dataset has a string of 0 or 0.0, it has to be appended to the new lists for being free. 

Then we just checked the results, by measuring the length of the new lists. Meaning that now we have **8.864** free apps in Android and **3.222** in iOS. 

After eliminating incorrect data, duplicate, non-english and non-free apps, we can say for sure that we've completed the purgation of the datasets and we're ready to analyze the valid data. 

We need to make our app profitable, for that we need to choose which apps will be more likely to attract more users, since our finantial gains come from the in-game adds. The more people playing our apps, the more posibilities they click on our adds. 

For that, we will develop our app for Google and iOS, which means to find a profile thaat is successful in both of them. The best way to find that successful profile will be determined by which genres are the most popular on both databases, as we it will give us more audience. 

In [18]:
print(android_header)
print('\n')
print(iOS_header)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


We will create frequency tables with the data from the column *'Genres'* in android and *'prime_genre'* in iOS, to determine the most popular genres. 

In [19]:
def freq_table(dataset, index):
    table = {}
    total = 0
    
    for row in dataset:
        total += 1
        value = row[index]
        if value in table:
            table[value] += 1
        else:
            table[value] = 1
    
    table_percentages = {}
    for key in table:
        percentage = (table[key] / total) * 100
        table_percentages[key] = percentage
    
    return table_percentages

def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])


Although lengthy, this is a brief description of what we've done in this code:

First off, we will create a function that generates frequency tables with percentages. That function will be named **freq_table** which has two imputs, *dataset* (a list of lists) and *integer* (an integer). Inside the function an empty dictionary named **table**, which will be our frequency table and the function *total* with a 0 value, which doesn't make sense for now, but it will.

Then we've looped the *row* in the *dataset* in order to count all the data.The loop will sum 1 to the *total*, that's why we needed it, then the *row* will go under the *value* variable. Then, we introduce an IF variable.

If *value* (which takes data from the row) exist in the dictionary table, will sum up 1. In other words, if something is repeated in any row, the loop will count it. If not (*else*) will stay as an unique value, as one. 

So far, we would have created a regular frequency table. But as we stated earlier on, we don't need integer or float numbers, but percentages. Not how many action games we have in the dataset, but which percentage they repressent. 

For that we create another empty dictionary named **table_percentages**, where we loop the *key* of the previous dictionary *table*. We have calculated the percentage by dividing that dictionary by the *total* and then multiply the result by **100**. Everything will be under the variable *percentage*.

Finally we wrap up everything by returning **table_percentages**.

Second step is to create another function that will sort those percentages in a descending order.

The **display_table** also takes as imput *dataset* and *index*. Our function *table* will take the data from the previous function. Then, we haven't created another new dictionary, but a list (*table_display*) and we used another variable. If the *key* is in *table*, we will transform the frequency table into a list of tuples (*key_val_as_tuple*) and append the results to the empty list. 

In the end we will sort the entries of the frequency table in descending order. 

In [20]:
display_table(iOS_free, 11)

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


We've tested our brand new function **display_table** for us to check out the results. 
As we can see, more than half of the apps in the iOS database are games (**58%**), winning by a landslide, while its closest competitor doesn't even reach an 8%, which is Entertainment. 

Correlation doesn't means causality, which means that nothing guarentees that most of the users on those apps are active users, they just might download the game, play for a few times and then abandon it. Developing an app for iOS wouldn't be a risky movement, but it doesn't ensures a large chunk of users. Although it's worth developing. 

It's also noticeable that the most prominemt genres of apps are leisure-based, just for killing time, such as Games, Entertainment, Social Networking... While the "practical" ones, like "Photo & Video", "Finance", or "Weather" are a minority. 

A good option in this case would be to focus our developments on those apps for "fun", as they are majority in the Apple store, although we take the risk of entering a saturated market where it's difficult to get noticed. 

In [21]:
display_table(android_free, 1)

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 

The "Category" in the Google playstore show us a very different picture. While the first category (Family) doubles the second (Game), the percentage between them isn't so gigantic (18%-9%). In general, the genres have a percentage much better distributed. Even the practical apps, such as "Tools", "Business", "Lifestyle" are near the top places. While we don't exactly know what "Family" category makes reference to, it wouldn't be a wild guess imagining that could be games for infants. 

Even though two "fun" categories copes the first two places, the results tells us that the audience of Android users is much more diverse and versatile, while the Apple much more niche. 

In [22]:
display_table(android_free, 9)

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.9250902527075

The "genres" category confirms us our hypothesis. In this case, even a practical category like "Tools" leads the percentage ranking, while others like "Education", "Business", "Productivity" or "Lifestyle" are among the first places, while the fun genres are much more isolated in the ranking. 

We could sum up that our approach for iOS and Android must be radically different if we have to develop an app, since the users are nothing alike. 

Our next step is try to figure out which genres have the most users. One good way to do it is by seeing the number of installs of each app. While the Android database has a column named *“Installs”*, the Apple database doesn’t have one, the best we can get for the total number of users of an app is a proxy. In this case we will use *rating_count_tot*, the total number of ratings that each app has. This could prove that their users have been active enough to rate the app, therefore have been active enough to do it.

In [23]:
genres_iOS = freq_table(iOS_free, -5)

for genre in genres_iOS:
    total = 0
    len_genre = 0
    for app in iOS_free:
        genre_app = app[-5]
        if genre_app == genre:
            n_ratings = float(app[5])
            total += n_ratings
            len_genre += 1
            
    avg_number_ratings = (total / len_genre)
    print(genre, ':', avg_number_ratings)
    

        


Social Networking : 71548.34905660378
Photo & Video : 28441.54375
Games : 22788.6696905016
Music : 57326.530303030304
Reference : 74942.11111111111
Health & Fitness : 23298.015384615384
Weather : 52279.892857142855
Utilities : 18684.456790123455
Travel : 28243.8
Shopping : 26919.690476190477
News : 21248.023255813954
Navigation : 86090.33333333333
Lifestyle : 16485.764705882353
Entertainment : 14029.830708661417
Food & Drink : 33333.92307692308
Sports : 23008.898550724636
Book : 39758.5
Finance : 31467.944444444445
Education : 7003.983050847458
Productivity : 21028.410714285714
Business : 7491.117647058823
Catalogs : 4004.0
Medical : 612.0


To calculate the average number of users based on the proxy of *rating_count_tot* we've done the following above:

We created a new variable, *genres_iOS* which uses our previous function *freq_table*, because we will need to display the results as a frequency table, which will use data from our iOS dataset.

Then we've looped the *genres* in our new variable *genres_iOS* and created two more with a value of 0, to be fullied later on, in this case *total* and *len_genre*. 

Once again we've used another loop, in this case looping the *apps* in our cleaned dataset *iOS_free*. That loop has the variable *genre_app* which contains the app genre from *iOS_free*, and then wrote a condition. If *genre_app* is the same as *genre*, then we convert the user ratings (the fifth column, starting from 0) to a float and under the *n_ratings* variable (for number of ratings), we sum that last one to the *total* and sum *len_genre*. That's why they were at zero at the beginning. 

Thanks to those sums we can divide one by another to finally know the average number of ratings and finally print the result. 



The results are quite surprising, the navigation apps are in the top spot with **86090** units, while the social media apps, like **Facebook**, **Instagram** or **Tik Tok** are in third place with **71548**, falling close to the *Reference* apps, with **74942**. While some more recreational apps, like **Photo**, **Games** or **Music** are way behind the podium, being surpassed by more practical options such as **Weather**. 

In hindsight this makes sense: While not every iOS user will play long enough to a game to leave a review and it's not that important to know if a game it's good enough, other apps, like a GPS system or a Weather forecast are targeted to a broader target audience and the reviews are a key aspect. Many users might consider valuable to leave a review for an unrealiable GPS to save other people from the issues they may had experienced. 

This leads us to the main reason to develop a "practical" app for iOS. Users tend to leave a huge amount of reviews, which if we do our app properly could gives us visibility in the best way possible.

For the second reason, we have to go back a little bit. The frequency table we created to display the more popular apps in the AppStore shown us that recreational apps are way more common, but it's harder to stand out among the crowd. While the practical apps seem to have more dedicated users.

In [24]:
display_table(iOS_free, 11)

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


A good bet could be to develop **Reference** apps. What are they? A [quick look on the Apple Store](https://apps.apple.com/gb/charts/iphone/reference-apps/6006) proves that they are related to language, translation and even religion. Another good option could be **Wheater** apps, since everyone checks it out at least briefly. **Health & Fitness** is another plausible choice since its an upwarding trend. We could say the same with **Book**, as reading habits are slowling turning into e-books, for its practicality and the option of reading many books for free. At last but not least, we can't discard the **Utilities** apps as they can bring us many users. 

While others like **Music**, or **Finance** doesn't seem a very good fit, since they are very specific and requires additional programming. An app like Spotify, would require to buy publication rights for the music, and **Finance** implies managing bank accounts, transfers and encrypted coding information that would require an additional amount of work for very little benefit. 

We'll try to do the same, but this time applied to the Google Play dataset. In this case we don't need to use a proxy column, since the fifth column on the Google dataset is the number of installs. 

Thanks to own function to display frequency tables we can see that the numbers are quite useful, but also quite vague, as they don't tell us exact quantities, but approximated.

We don't know if an app with 1.000.000+ installs has 1.500.000, 1.600.000 and so on. Although we could have a vague idea about which apps possess the biggest amount of installations, we don't have that precise number.

In order to not complicate our investigation further, we will keep it simple, as we will take for granted that 1.000+ installs repressents 1.000 installs, 5.000.000+ will repressent 5.000.000+ and so on. 

In [25]:
display_table(android_free, 5)

1,000,000+ : 15.726534296028879
100,000+ : 11.552346570397113
10,000,000+ : 10.548285198555957
10,000+ : 10.198555956678701
1,000+ : 8.393501805054152
100+ : 6.915613718411552
5,000,000+ : 6.825361010830325
500,000+ : 5.561823104693141
50,000+ : 4.7721119133574
5,000+ : 4.512635379061372
10+ : 3.5424187725631766
500+ : 3.2490974729241873
50,000,000+ : 2.3014440433213
100,000,000+ : 2.1322202166064983
50+ : 1.917870036101083
5+ : 0.78971119133574
1+ : 0.5076714801444043
500,000,000+ : 0.2707581227436823
1,000,000,000+ : 0.22563176895306858
0+ : 0.04512635379061372
0 : 0.01128158844765343


In [26]:
categories_android = freq_table(android_free, 1)

for category in categories_android:
    total = 0
    len_category = 0
    for app in android_free:
        category_app = app[1]
        if category_app == category:
            n_installs = app[5]
            n_installs = n_installs.replace('+', '')
            n_installs = n_installs.replace(',', '')
            total += float(n_installs)
            len_category += 1
            
    avg_number_installs = total / len_category
    print(category, ':', avg_number_installs)

ART_AND_DESIGN : 1986335.0877192982
AUTO_AND_VEHICLES : 647317.8170731707
BEAUTY : 513151.88679245283
BOOKS_AND_REFERENCE : 8767811.894736841
BUSINESS : 1712290.1474201474
COMICS : 817657.2727272727
COMMUNICATION : 38456119.167247385
DATING : 854028.8303030303
EDUCATION : 1833495.145631068
ENTERTAINMENT : 11640705.88235294
EVENTS : 253542.22222222222
FINANCE : 1387692.475609756
FOOD_AND_DRINK : 1924897.7363636363
HEALTH_AND_FITNESS : 4188821.9853479853
HOUSE_AND_HOME : 1331540.5616438356
LIBRARIES_AND_DEMO : 638503.734939759
LIFESTYLE : 1437816.2687861272
GAME : 15588015.603248259
FAMILY : 3695641.8198090694
MEDICAL : 120550.61980830671
SOCIAL : 23253652.127118643
SHOPPING : 7036877.311557789
PHOTOGRAPHY : 17840110.40229885
SPORTS : 3638640.1428571427
TRAVEL_AND_LOCAL : 13984077.710144928
TOOLS : 10801391.298666667
PERSONALIZATION : 5201482.6122448975
PRODUCTIVITY : 16787331.344927534
PARENTING : 542603.6206896552
WEATHER : 5074486.197183099
VIDEO_PLAYERS : 24727872.452830188
NEWS_AND_

We've done a similar calculation to the previous iOS dataset. In this occasion it was about to decipher how many installs we have per genre. And yes, we've displayed it as a frequency table. 

For that we've created a variable named **categories_android** where we will use the **freq_table** function with the column *'Category'* from android_free. 

Then looped the iteration *category* within our new variable **categories_android** and two new variables, *total* and *len_category* which we will leave empty. As we did with the iOS database, we will use again a nested loop for the apps in android_free, and created the variable *category_app* for the *'Category'* column. But here's when it turns tricky. We're working with strings, like 100,000+, and to perform computations we need to convert those to floats. But if we would that directly, due to the special symbols like ',' or '+', some kind of error would surface. For that we had to use the method **str.replace**, where we've subtituted those special symbols by nothing (using ''), going from 100,000+ to 100000, which is easier to convert to float and then, we added that number of installs to the total, and finally incremented len_category by 1. So that both variables aren't empty anymore

Once again, as we did with iOS, we've divided *total* by *len_category* to obtain the average number of installs on each category and printed the *category* and the *average number of installs* in order to have that frequency table with all the data. 

As we can see, the *communication* apps are the most common with **38456119**, possibly because under this category titans such as Whatsapp or Skype can be found here, something that would probably raises the mean considerably. We could do a loop to see if that's right.

In [27]:
for app in android_free:
    if app[1] == 'COMMUNICATION' and (app[5] == '1,000,000,000+' or app[5] == '500,000,000+' or app[5] == '100,000,000+'):
        print(app[0], ':', app[5])
        

WhatsApp Messenger : 1,000,000,000+
imo beta free calls and text : 100,000,000+
Android Messages : 100,000,000+
Google Duo - High Quality Video Calls : 500,000,000+
Messenger – Text and Video Chat for Free : 1,000,000,000+
imo free video calls and chat : 500,000,000+
Skype - free IM & video calls : 1,000,000,000+
Who : 100,000,000+
GO SMS Pro - Messenger, Free Themes, Emoji : 100,000,000+
LINE: Free Calls & Messages : 500,000,000+
Google Chrome: Fast & Secure : 1,000,000,000+
Firefox Browser fast & private : 100,000,000+
UC Browser - Fast Download Private & Secure : 500,000,000+
Gmail : 1,000,000,000+
Hangouts : 1,000,000,000+
Messenger Lite: Free Calls & Messages : 100,000,000+
Kik : 100,000,000+
KakaoTalk: Free Calls & Text : 100,000,000+
Opera Mini - fast web browser : 100,000,000+
Opera Browser: Fast and Secure : 100,000,000+
Telegram : 100,000,000+
Truecaller: Caller ID, SMS spam blocking & Dialer : 100,000,000+
UC Browser Mini -Tiny Fast Private & Secure : 100,000,000+
Viber Mess

In fact, those apps raises the mean pretty much, as we guessed. If we eliminate those from the ecuation, the result would probably be quite different. 

In [28]:
less_than_100_m = []

for app in android_free:
    n_installs = app[5]
    n_installs = n_installs.replace('+', '')
    n_installs = n_installs.replace(',', '')
    if (app[1] == 'COMMUNICATION') and (float(n_installs) < 100000000):
        less_than_100_m.append(float(n_installs))
        
sum(less_than_100_m) / len(less_than_100_m)

3603485.3884615386

After eliminating the apps with more than **100.000.000**, the mean has dropped down from **38.456.119** to just **3.603.485**

Going back to our original frequency table with the most popular categories, we could tell the same story from more popular categories: A few apps in the top that have numbers so immense that they distort the real numbers, like *Game* with **15.588.015**, or *Social* with **23.253.652**. Would it be intelligent to develop an app in those categories? The number of users are attractive enough to feel tempted by that possibility, but like we said in the iOS case, it possesses the great risk of going unnoticed on an already saturated market.

For that, we're keeping the same strategy and going for categories not so populated and the giants aren't so strong. *Books and reference* with **8.767.811** or *Health and fitness* with **4.188.821** could be good alternatives.

In [29]:
for app in android_free:
    if app[1] == 'HEALTH_AND_FITNESS' and (app[5] == '100,000,000+' or app[5] == '10,000,000+' or app[5] == '1,000,000+' or app[5] == '100,000+'):
        print(app[0], ':', app[5])

Pedometer - Step Counter Free & Calorie Burner : 1,000,000+
Six Pack in 30 Days - Abs Workout : 10,000,000+
Lose Weight in 30 Days : 10,000,000+
Pedometer : 10,000,000+
LG Health : 10,000,000+
Step Counter - Pedometer Free & Calorie Counter : 10,000,000+
Pedometer, Step Counter & Weight Loss Tracker App : 10,000,000+
Sportractive GPS Running Cycling Distance Tracker : 1,000,000+
30 Day Fitness Challenge - Workout at Home : 10,000,000+
Home Workout for Men - Bodybuilding : 1,000,000+
Fat Burning Workout - Home Weight lose : 100,000+
Walking for Weight Loss - Walk Tracker : 100,000+
Sleep Sounds : 1,000,000+
Fitbit : 10,000,000+
Abs Training-Burn belly fat : 100,000+
Calorie Counter - EasyFit free : 1,000,000+
Garmin Connect™ : 10,000,000+
Bike Computer - GPS Cycling Tracker : 1,000,000+
Six Packs for Man–Body Building with No Equipment : 100,000+
Running Distance Tracker + : 1,000,000+
The TK-App - everything under control : 100,000+
Runkeeper - GPS Track Run Walk : 10,000,000+
Walking:

It would be worth to giving a fight on this field. While there are some huge apps with millions of users, they are not as big as we could find in other niche categories, while there's many apps with the same number of users with not-so-intimidating numbers. We would need to implement some novelties, which would require more specific research on the most popular apps, but at least we know that it isn't impossible to build a decent quantity of users. 

In [30]:
for app in android_free:
    if app[1] == 'ENTERTAINMENT' and (app[5] == '100,000,000+' or app[5] == '10,000,000+' or app[5] == '1,000,000+' or app[5] == '100,000+'):
        print(app[0], ':', app[5])

Complete Spanish Movies : 1,000,000+
Pluto TV - It’s Free TV : 1,000,000+
Mobile TV : 10,000,000+
Motorola Spotlight Player™ : 10,000,000+
Hotstar : 100,000,000+
The green alien dance : 1,000,000+
StarTimes - Live International Champions Cup : 1,000,000+
Cinematic Cinematic : 1,000,000+
MEGOGO - Cinema and TV : 10,000,000+
Talking Angela : 100,000,000+
ivi - movies and TV shows in HD : 10,000,000+
Radio Javan : 1,000,000+
🔥 Football Wallpapers 4K | Full HD Backgrounds 😍 : 1,000,000+
Movies by Flixster, with Rotten Tomatoes : 10,000,000+
Low Poly – Puzzle art game : 1,000,000+
BBC Media Player : 10,000,000+
Adult Glitter Color by Number Book - Sandbox Pages : 1,000,000+
IMDb Movies & TV : 100,000,000+
Ziggo GO : 1,000,000+
Talking Ben the Dog : 100,000,000+
Trailer Addict Movie Trailers : 100,000+
Cinemark Theatres : 1,000,000+
Regal Cinemas : 1,000,000+
Fandango Movies - Times + Tickets : 10,000,000+
Marcus Theatres : 100,000+
Harkins Theatres : 100,000+
AMC Theatres : 1,000,000+
Cooki

Others, like Entertainment could be a good option, but we would need to gather audiovisual rights if we want to show movies, leaving us with just pictures. Which isn't worthy. 

In [31]:
for app in android_free:
    if app[1] == 'BOOKS_AND_REFERENCE' and (app[5] == '1,000,000,000+' or app[5] == '10,000,000+' or app[5] == '1,000,000+' or app[5] == '100,000+'):
        print(app[0], ':', app[5])

Download free book with green book : 100,000+
Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Free Panda Radio Music : 100,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
Google Play Books : 1,000,000,000+
Offline English Dictionary : 100,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
ReadEra – free ebook reader : 1,000,000+
Litnet - E-books : 100,000+
eBoox: book reader fb2 epub zip : 1,000,000+
All Maths Formulas : 1,000,000+
HTC Help : 10,000,000+
English translation from Bengali : 100,000+
Pdf Book Download - Read Pdf Book : 100,000+
Free Book Reader : 100,000+
Moon+ Reader : 10,000,000+
English-Myanmar Dictionary : 1,000,000+
Golden Dictionary (EN-AR) : 1,000,000+
All Language Translator Free : 1,000,000+
URBANO V 02 instruction manual : 100,000+
Aab e Hayat Full Novel : 100,000+
Aldiko Book Reader : 10,000,000+
V Made : 100,000+
Guide (for X-MEN) : 100,000+
Al-Quran (Free) : 10,00

Following the example of iOS and "Reference", we can take the road of developing literacy related apps. While the field is dominated by translators and dictionaries (Spanish English Translator, English Hindi Dictionary, Dictionary...) We could think about developing an additional feature, like audiobooks, a marker for the most important part of each book, etc. That could attract potential new users. 

Our final thoughts: We should invest, on both Android and iOS, on categories where there's a decent amount of users, not so niche, but avoiding those who has gigantic apps, such as Disney+ or Whatsapp. Like we said, **books** and **health** could be two fields where we could stand out. 